core/ibft.go

```diff
@@ -0,0 +1,1134 @@
+		case <-ctx.Done():
+			return
+		case round := <-sub.SubCh:
+			proposal := i.handlePrePrepare(&proto.View{Height: height, Round: round})
+			if proposal == nil {
+				continue
+			}
+
+			// Extract the proposal
+			i.signalNewProposal(
+				ctx,
+				newProposalEvent{proposal, round},
+			)
+
+			return
+		}
+	}
+}
+
+// watchForRoundChangeCertificates is a routine that waits
+// for future valid Round Change Certificates that could
+// trigger a round hop
+func (i *IBFT) watchForRoundChangeCertificates(ctx context.Context) {
+	defer i.wg.Done()
+
+	var (
+		view   = i.state.getView()
+		height = view.Height
+		round  = view.Round
+		quorum = i.backend.Quorum(height)
+
+		sub = i.messages.Subscribe(messages.SubscriptionDetails{
+			MessageType: proto.MessageType_ROUND_CHANGE,
+			View: &proto.View{
+				Height: height,
+				Round:  round + 1, // only for higher rounds
+			},
+			NumMessages: 1,
+			HasMinRound: true,
+		})
+	)
+
+	defer i.messages.Unsubscribe(sub.ID)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case round := <-sub.SubCh:
+			rcc := i.handleRoundChangeMessage(
+				&proto.View{
+					Height: height,
+					Round:  round,
+				},
+				quorum,
+			)
+			if rcc == nil {
+				continue
+			}
+
+			//	we received a valid RCC for a higher round
+			i.signalNewRCC(ctx, round)
+
+			return
+		}
+	}
+}
+
+// RunSequence runs the IBFT sequence for the specified height
+func (i *IBFT) RunSequence(ctx context.Context, h uint64) {
+	// Set the starting state data
+	i.state.clear(h)
+	i.messages.PruneByHeight(h)
+
+	i.log.Info("sequence started", "num", h)
+	defer i.log.Info("sequence done", "num", h)
+
+	for {
+		i.wg.Add(4)
+
+		view := i.state.getView()
+
+		i.log.Info("round started", "num", view.Round)
+
+		currentRound := view.Round
+		ctxRound, cancelRound := context.WithCancel(ctx)
+
+		// Start the round timer worker
+		go i.startRoundTimer(ctxRound, currentRound)
+
+		//	Jump round on proposals from higher rounds
+		go i.watchForFutureProposal(ctxRound)
+
+		//	Jump round on certificates
+		go i.watchForRoundChangeCertificates(ctxRound)
+
+		// Start the state machine worker
+		go i.runRound(ctxRound)
+
+		teardown := func() {
+			cancelRound()
+			i.wg.Wait()
+		}
+
+		select {
+		case ev := <-i.newProposal:
+			teardown()
+			i.log.Info("received future proposal", "round", ev.round)
+
+			i.moveToNewRound(ev.round)
+			i.acceptProposal(ev.proposalMessage)
+			i.state.setRoundStarted(true)
+		case round := <-i.roundCertificate:
+			teardown()
+			i.log.Info("received future RCC", "round", round)
+
+			i.moveToNewRound(round)
+		case <-i.roundExpired:
+			teardown()
+			i.log.Info("round timeout expired", "round", currentRound)
+
+			newRound := currentRound + 1
+			i.moveToNewRound(newRound)
+
+			i.transport.Multicast(
+				i.backend.BuildRoundChangeMessage(
+					i.state.getLatestPreparedProposedBlock(),
+					i.state.getLatestPC(),
+					&proto.View{
+						Height: h,
+						Round:  newRound,
+					},
+				),
+			)
```

Comment:
Why not create a method that sends round change method?

Reply:
Simplified it for other types of messages as well:

[80ba85c](https://github.com/0xPolygon/go-ibft/pull/21/commits/80ba85c356c35f186e54f450797e658d45ee602e)